mlflow/projects/docker.py

```diff
@@ -26,15 +26,12 @@
 
 def validate_docker_installation():
     """
-    Verify if Docker is installed on host machine.
+    Verify if Docker is installed and running on host machine.
     """
-    try:
-        docker_path = "docker"
-        process._exec_cmd([docker_path, "--help"], throw_on_error=False)
-    except EnvironmentError:
+    if shutil.which("docker") is None:
```

Comment:
quick test to check what `shutil.which("docker")` returns:

```
>>> docker run --rm python:3.7 python -c "import shutil; print(shutil.which('docker'))"
None

>>> python -c "import shutil; print(shutil.which('docker'))"
/usr/bin/docker
```

Reply:
this returns the existence of the docker binary, not whether it's running.

```python
client = docker.from_env()
```

this will throw error when docker not running.

Reply:
@stobias123 Thanks for the comment! You're right. I think we need to run a docker command (e.g. docker ps) that fails when docker is not running.

Reply:
Test on my machine:

```
# When docker executable is available and docker daemon is running

% docker ps
CONTAINER ID   IMAGE      COMMAND                  CREATED       STATUS         PORTS      NAMES
9a4f895cbd66   postgres   "docker-entrypoint.s…"   7 weeks ago   Up 4 minutes   5432/tcp   db-postgresql-1
                                                                                                                                                
(.venv) @ mlflow [fix-5851] -- INSERT --
% python3 -c "from mlflow.projects.docker import validate_docker_installation; validate_docker_installation()"
```

```
# When docker executable is available but docker daemon is not running

% docker ps
Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
                                                                                                                                                
(.venv) @ mlflow [fix-5851] -- INSERT --
% python3 -c "from mlflow.projects.docker import validate_docker_installation; validate_docker_installation()"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/Users/harutakakawamura/Desktop/repositories/mlflow/mlflow/projects/docker.py", line 42, in validate_docker_installation
    raise ExecutionException(
mlflow.exceptions.ExecutionException: Ran `docker ps` to ensure docker daemon is running but it failed: Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
```

```
# docker executable is unavailable
                                                                                                                                                
(.venv) @ mlflow [fix-5851] -- INSERT --
% docker run --rm -w /app -v $(pwd):/app python:3.7 bash -c 'pip install -e . && python3 -c "from mlflow.projects.docker import validate_docker_installation; validate_docker_installation()"'
...
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/app/mlflow/projects/docker.py", line 33, in validate_docker_installation
    "Could not find Docker executable. "
mlflow.exceptions.ExecutionException: Could not find Docker executable. Ensure Docker is installed as per the instructions at https://docs.docker.com/install/overview/.
```